lib/transitions.bzl

```
@@ -42,3 +44,60 @@ platform_transition_filegroup = rule(
     },
     doc = "Transitions the srcs to use the provided platform. The filegroup will contain artifacts for the target platform.",
 )
+
+def _platform_transition_binary_impl(ctx):
+    # We need to forward the DefaultInfo provider from the underlying rule.
+    # Unfortunately, we can't do this directly, because Bazel requires that the executable to run
+    # is actually generated by this rule, so we need to symlink to it, and generate a synthetic
+    # forwarding DefaultInfo.
+
+    result = []
+    binary = ctx.attr.binary[0]
+
+    default_info = binary[DefaultInfo]
+
+    new_executable = None
+    files = default_info.files
+    original_executable = default_info.files_to_run.executable
+    runfiles = default_info.default_runfiles
+    if original_executable:
+        new_executable_name = ctx.attr.basename if ctx.attr.basename else original_executable.basename
+
+        # In order for the symlink to have the same basename as the original
+        # executable (important in the case of proto plugins), put it in a
+        # subdirectory named after the label to prevent collisions.
+        new_executable = ctx.actions.declare_file(paths.join(ctx.label.name, new_executable_name))
+        ctx.actions.symlink(
+            output = new_executable,
+            target_file = original_executable,
+            is_executable = True,
+        )
+        files = depset(direct = [new_executable])
```

Why doesn't this include the files?

---

@illicitonion Was there a reason for not including the files from DefaultInfo in your original snippet?

---

Either it's an accidental omission, or it's forbidden. Feel free to give it a try! If it works, great! If it doesn't, I suspect there may be a restriction where `DefaultInfo.files` is only allowed to contain files generated by the current rule.

---

Right, makes sense that if the output it just a transitioned binary then that's all we should have in `files`.

It looks like if the `binary` is not executable, we just forward the default info, but passing in a non executable label isn't a valid case, right? I added a fixup commit to just fail early instead.

---

Look at my implementation above in https://github.com/aspect-build/bazel-lib/blob/2c66c1c778d1050bf147a9a65fcc6f53ee47f938/lib/transitions.bzl#L15-L24 for an example of how I did it.

Only forwarding the symlink to the binary may work in most cases, but you would miss files, e.g., when you transition it to add to a tarball. It won't necessarily pick the runfiles. This can include shared objects required by the binary.

---

Added the binary's files as transitive deps.

---

Perfect, thank you!

---

Ooh, interesting... I just tried to apply this everywhere I was using this pattern before, and found that as a side-effect of this change, we can't use a `platform_transition_binary` as a value of an attribute which is set to `allow_single_file = True`, which makes it a little less useful...

---

@illicitonion interesting. Does Bazel accept the original binary you are trying to transition as a single file? (I know it won't run, but at least pass the analysis phase)

---

> @illicitonion interesting. Does Bazel accept the original binary you are trying to transition as a single file? (I know it won't run, but at least pass the analysis phase)

Yeah, the original `go_binary` in my example passes.

When using `platform_transition_binary`, the issue is because `DefaultInfo.files` on the `platform_transition_binary` target has two files. If I remove the `transitive` arg, everything starts working again.

---

This might be similar to this bug: https://github.com/bazelbuild/bazel/pull/16381.